### ISSUE TITLE
Fog OpenStack keystone v3 caching is non deterministic

### ISSUE THREAD

**Issue by Ladas** [fog/fog#3812]
brought by:
https://github.com/fog/fog/pull/3738

My biggest problem is, that the behavior is non deterministic now, which is not good for lib I think. So for ManageIQ CI, which is doing a lot of request per one recorded VCR, it will always pass the 30 second interval, always having variable amount of requests cached. Then of course the recorded VCR is much quicker, so that is another issue, it would require having sleep there.

So, this makes it impossible to get deterministic behavior. When Invoking the same methods, I am getting basically random API requests fired. And that always breaks our CI.

My proposed solution:
1. Having it configurable, disabled caching by default, we would pass e.g. :cache_ttl => 30 as options for auth request, then other services could take the settings from there, e.g. for keystone project list, etc.
So it would be Fog::Compute.new({..., :cache_ttl => 30 })

2. Revert the caching here and rather rely on application specific implementation of caching, using redis, memcache, etc. But unless we will just have http layer caching, we would need to expose the Auth object.

So e.g. something like this:
auth = Fog::Auth.new({
    :provider            => 'openstack',                                      # OpenStack Fog provider
    :openstack_auth_url  => 'http://KEYSTONE_HOST:KEYSTONE_PORT/v2.0/tokens', # OpenStack Keystone endpoint
    :openstack_username  => OPEN_STACK_USER,                                  # Your OpenStack Username
    :openstack_tenant    => OPEN_STACK_TENANT,                                # Your tenant id
    :openstack_api_key   => OPEN_STACK_PASSWORD,                              # Your OpenStack Password
    :connection_options  => {}                                                # Optional
})

then:

Fog::Compute.new(auth_object => auth)
Fog::Image.new(auth_object => auth)

I think these would not need to do API request then, if we will fetch auth token and keystone catalog into the auth object. And the Auth object could support rescoping maybe? Which we use for changing a tenant scope e.g. Fog::Image.new(auth_object => auth, openstack_tenant => 'project1'), which would just fire one api request for rescoping using existing token.

**Comment by Ladas** [comment]
@geemus @dhague @TerryHowe Any suggestions? 

This kinda blocks me from using the new Fog version. ( well I would have to remove my keystone v3 tests from CI :-) )

**Comment by dhague** [comment]
Solution (1) makes sense to me - cache_ttl would be zero by default, and would disable caching. That would involve minimal coding, resolve the issue you are having, and still allow our Ruby-on-Rails application to behave reasonably quickly by turning on per-request caching with a TTL of 30 seconds.

**Comment by dhague** [comment]
Solution (2) also sounds good, but would require rather more coding :-P

**Comment by Ladas** [comment]
@dhague I agree :-)

Do you think you will have a time for the for Solution (1), or should I do that?

**Comment by dhague** [comment]
I won't have the time this week, but should be OK to do it next week if that works for you?

**Comment by Ladas** [comment]
@dhague if you will have time for it next week, that would be awesome, thank you. :-)

**Comment by geemus** [comment]
Sounds like starting with solution 1 will release the pressure and then maybe we can revisit and think more about 2 at some point in the future.

**Comment by Ladas** [comment]
@geemus yeah, I think I might tackle the solution 2 in the future, as it will allow optimization of API calls
